core/kazoo_services/src/kz_services.erl

```diff
@@ -1184,6 +1185,92 @@ save_services_jobj(Services, ProposedJObj) ->
+%% @doc
+%% @end
+%%------------------------------------------------------------------------------
+-spec maybe_audit_reseller(services(), kz_json:object(), kz_json:object()) -> 'ok'.
+maybe_audit_reseller(Services, CurrentServicesJObj, ProposedServicesJObj)  ->
+    case has_quantity_changes(CurrentServicesJObj, ProposedServicesJObj) of
+        'false' -> 'ok';
+        'true' ->
+          audit_reseller(Services, CurrentServicesJObj, ProposedServicesJObj)
+   end.
+
+%%------------------------------------------------------------------------------
+%% @doc
+%% @end
+%%------------------------------------------------------------------------------
+-spec audit_reseller(services(), kz_json:object(), kz_json:object()) -> 'ok'.
+audit_reseller(Services, CurrentServicesJObj, ProposedServicesJObj) ->
+    {'ok', MasterAccountId} = kapps_util:get_master_account_id(),
+    AccountId = account_id(Services),
+    ResellerId = kz_services_reseller:get_id(AccountId),
+
+    case kz_services_reseller:is_reseller(AccountId)
+         andalso MasterAccountId =/= AccountId of
+        % don't cascade audit to Master Acct
+        'false' -> 'ok';
+        'true' ->
+            lager:debug("creating account ~s reseller audit for ~s", [AccountId, ResellerId]),
+            JObj = kz_json:from_list_recursive([{<<"id">>, kazoo_yodb_util:yodb_id()}
+                                               ,{<<"audit">>
+                                                ,generate_diff_obj(CurrentServicesJObj, ProposedServicesJObj)
+                                                }
+                                               ]),
+            Doc = kz_doc:update_pvt_parameters(JObj, kazoo_yodb:get_yodb(AccountId)),
+            _ = kazoo_yodb:save_doc(AccountId, Doc),
+            'ok'
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc run a diff between services jobs to determine if cascade quantities
+%% changed.
+%% @end
+%%------------------------------------------------------------------------------
+-spec has_quantity_changes(kz_json:object(), kz_json:object()) -> boolean().
+has_quantity_changes(CurrentServicesJObj, ProposedServicesJObj) ->
+    CascadesQty = [<<"quantities">>],
+    PQ = kz_json:get_value(CascadesQty, ProposedServicesJObj),
+    CQ = kz_json:get_value(CascadesQty, CurrentServicesJObj),
+    Diff = kz_json:diff(CQ, PQ),
+    case kz_json:is_empty(Diff) of
+        'true' -> 'false';
+        _ -> 'true'
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc extracts diff of services and returns audit changes JObj
+%% @end
+%%------------------------------------------------------------------------------
+-spec generate_diff_obj(kz_json:object(), kz_json:object()) -> kz_json:object().
+generate_diff_obj(CurrentServicesJObj, ProposedServicesJObj) ->
+    Key = [<<"quantities">>],
+    PQ = kz_json:get_value(Key, ProposedServicesJObj),
```

Comment:
Can you use a typed getter here? `kz_json:get_json_value(Key, ProposedServicesJObj)` will help know what PQ should be later on.